vm/src/air_public_input.rs

```diff
@@ -0,0 +1,122 @@
+use std::collections::HashMap;
+
+use felt::Felt252;
+use serde::Serialize;
+use thiserror::Error;
+
+use crate::{
+    types::layout::CairoLayout,
+    vm::{
+        errors::{trace_errors::TraceError, vm_errors::VirtualMachineError},
+        trace::trace_entry::TraceEntry,
+    },
+};
+
+#[derive(Serialize, Debug)]
+pub struct PublicMemoryEntry {
+    address: usize,
+    page: usize,
+    #[serde(serialize_with = "mem_value_serde::serialize")]
+    value: Option<Felt252>,
+}
+
+mod mem_value_serde {
+    use super::*;
+    use serde::Serializer;
+
+    pub(crate) fn serialize<S: Serializer>(
+        value: &Option<Felt252>,
+        serializer: S,
+    ) -> Result<S::Ok, S::Error> {
+        if let Some(value) = value {
+            serializer.serialize_str(&format!("0x{}", value.to_str_radix(16)))
+        } else {
+            serializer.serialize_none()
+        }
+    }
+}
+
+#[derive(Serialize, Debug)]
+pub struct PublicInput<'a> {
+    layout: &'a str,
+    layout_params: Option<&'a CairoLayout>,
+    rc_min: Option<isize>,
+    rc_max: Option<isize>,
+    n_steps: usize,
+    memory_segments: HashMap<&'a str, (usize, Option<usize>)>,
+    public_memory: Vec<PublicMemoryEntry>,
+}
+
+impl<'a> PublicInput<'a> {
+    pub fn new(
+        memory: &[Option<Felt252>],
+        layout: &'a str,
+        dyn_layout_params: Option<&'a CairoLayout>,
+        public_memory_addresses: &[(usize, &usize)],
```

Comment:
I find this type strange. Why is the second element of the tuple a reference?

Reply:
To avoid a `.clone()` call, but if you prefer I can make that call and change this (as I mentioned in another review)

Reply:
I was also cloning it for creating the `PublicInput` struct, I think I was going to remove the reference but forgot about it. Thank you! done.